boot/bootstate20.go

```diff
@@ -46,10 +48,37 @@ func newBootState20(typ snap.Type, dev snap.Device) bootState {
 	}
 }
 
+// modeenvMu is used to protect sections doing:
+//  * read moddeenv/modify it(/reseal from it)
+//  * write modeenv/seal from it
+// while we might want to release the global state lock as seal/reseal are slow
+// (see Unlocker for that)
+var (
+	modeenvMu     sync.Mutex
+	modeenvLocked int32
+)
+
+func modeenvLock() {
+	modeenvMu.Lock()
+	atomic.AddInt32(&modeenvLocked, 1)
+}
+
+func modeenvUnlock() {
+	atomic.AddInt32(&modeenvLocked, -1)
+	modeenvMu.Unlock()
+}
+
+func isModeeenvLocked() bool {
+	return atomic.LoadInt32(&modeenvLocked) == 1
+}
+
 func loadModeenv() (*Modeenv, error) {
+	if !isModeeenvLocked() {
```

Comment:
As discussed in the standup I mostly added this check to be sure through the unit tests, that I added locking as early as possible in all the relevant functions, but this is robust only as long as future code in boot and outside follows the current patterns. As I mentioned this should be good enough to unblock us but after this we should/could consider:

* stopping code outside of boot to write and read in-full the moedeenv, by unexporting the related methods and adding any necessary more specific helpers, otoh we will need some work to keep supporting the tests that need to write these
* having checks about the locks in the actual read/write in modeenv.go instead of one level up
* as you suggested adding some kind env_revno to the modeenv and checking it for conflicts when writing